# Manual Feature Selection

The simplest method of selecting features that offer predictive power to our target variable is to use our external and domain knowledge to select our features with confidence. In this notebook, we will recap simple methods that you've seen throughout the course, and recap when it appropriate to use them.

These manual methods will become more and more redundant as you explore more advanced techniques, but they sevre as a useful way to explore your features manually.

### Import Basic Packages & Data

In [21]:
#Load the required libraries

# Data manipulation libraries
import pandas as pd
import numpy as np

# Data visualisation libraries
import matplotlib.pyplot as plt
import seaborn as sns

In [22]:
# Import data to a pandas dataframe
df_cars = pd.read_csv('indian cars dataset incnulls.csv')
df_cars.head()

# We will be working with the Indian cars dataset. The target variable in this instance is the starting price of a vehicle.

ending_price  starting_price  reviews_count  max_torque_nm  max_torque_rpm  \
0        583000          399000             51           89.0            3500   
1       1396000          799000             86          136.8            4400   
2       1603000         1353000            242          300.0            2800   
3       2458000         1318000            313          450.0            2800   
4       2390000         1199000            107          400.0            2750   

   max_power_bhp  max_power_rp  fuel_tank_capacity  no_cylinder  rating  \
0          65.71          5500                27.0            3     4.5   
1         101.65          6000                48.0            4     4.5   
2         130.00          3750                57.0            4     4.5   
3         182.38          3500                60.0            4     4.5   
4         172.45          3500                57.0            4     4.5   

   seating_capacity fuel_type  engine_displacement transmission_type  \
0               5.0    Petrol                  998         Automatic   
1               5.0    Petrol                 1462         Automatic   
2               4.0    Diesel                 2184         Automatic   
3               7.0    Diesel                 2198         Automatic   
4               7.0    Diesel                 2198         Automatic   

   body_type            car_name  
0  Hatchback     Maruti Alto K10  
1        SUV       Maruti Brezza  
2        SUV       Mahindra Thar  
3        SUV     Mahindra XUV700  
4        SUV  Mahindra Scorpio-N

### Removing Features Using Domain Knowledge

Most importantly, we can use our domain or situation knowledge of the scenario we are working with to make conclusions about which features to select.

- Some data may simply be unavailable in a real world scenario.
- Some data might be available, but deemed too expensive to collect by the business. 
- Some features may have a known scientific or commonly known connection.

**In this case** we are trying to predict the starting price of each auction. We therefore know that the ending price will not be available to help us with the prediction (The ending price cannot exist if the starting price is still unknown). For this reason, the feature is irrelevant when predicting our target variable and can be removed.

In [23]:
# Removing irrelevant feature
df_cars = df_cars.drop(columns=['ending_price'])
df_cars

starting_price  reviews_count  max_torque_nm  max_torque_rpm  \
0            399000             51           89.0            3500   
1            799000             86          136.8            4400   
2           1353000            242          300.0            2800   
3           1318000            313          450.0            2800   
4           1199000            107          400.0            2750   
..              ...            ...            ...             ...   
198          659000             35          500.0            5250   
199         1041000              3          400.0            4400   
200         1615000              2          350.0            2500   
201        21700000              9          800.0            4500   
202          391000              4           69.0            3500   

     max_power_bhp  max_power_rp  fuel_tank_capacity  no_cylinder  rating  \
0            65.71          5500                27.0            3     4.5   
1           101.65          6000                48.0            4     4.5   
2           130.00          3750                57.0            4     4.5   
3           182.38          3500                60.0            4     4.5   
4           172.45          3500                57.0            4     4.5   
..             ...           ...                 ...          ...     ...   
198         415.71          6750                 0.0            4     4.5   
199         254.79          5000                59.0            4     4.5   
200         167.67          3750                60.0            4     4.5   
201         591.39          6000                85.0            8     3.5   
202          47.33          6000                35.0            3     4.5   

     seating_capacity fuel_type  engine_displacement transmission_type  \
0                 5.0    Petrol                  998         Automatic   
1                 5.0    Petrol                 1462         Automatic   
2                 4.0    Diesel                 2184         Automatic   
3                 7.0    Diesel                 2198         Automatic   
4                 7.0    Diesel                 2198         Automatic   
..                ...       ...                  ...               ...   
198               5.0    Petrol                 1991         Automatic   
199               5.0    Petrol                 1998         Automatic   
200               7.0    Diesel                 1956            Manual   
201               5.0    Petrol                 3998         Automatic   
202               5.0    Petrol                  796            Manual   

     body_type                     car_name  
0    Hatchback              Maruti Alto K10  
1          SUV                Maruti Brezza  
2          SUV                Mahindra Thar  
3          SUV              Mahindra XUV700  
4          SUV           Mahindra Scorpio-N  
..         ...                          ...  
198  Hatchback     Mercedes-Benz AMG A 45 S  
199      Sedan  BMW 3 Series Gran Limousine  
200        SUV               MG Hector Plus  
201        SUV                   Audi RS Q8  
202  Hatchback         Maruti Alto 800 tour  

[203 rows x 15 columns]

### Removing Features Based on Logical Reasoning

When we look at our dataset, we can use our understanding of our data to make some obvious decisions about predictive power. 

For example, we can say that the name of the car is not going to be helpful when predicting the starting price of a car. There are 199 distinct car names out of 203 rows of data, which would mean that there is hardly any predictive power in that feature. 

Inversely, if there is a feature in our dataset that has only one unique value, then there is no reason to keep that feature as that does not provide any predictive power as well.

In [24]:
# Use the unique function to explore how many unique values in each column.
df_cars.nunique()

starting_price         190
reviews_count           97
max_torque_nm           82
max_torque_rpm          50
max_power_bhp          143
max_power_rp            36
fuel_tank_capacity      47
no_cylinder              9
rating                   5
seating_capacity         6
fuel_type                4
engine_displacement     76
transmission_type        2
body_type               11
car_name               199
dtype: int64

With this knowledge, we will choose to drop the car name feature from our data.

In [25]:
df_cars = df_cars.drop(columns=['car_name'])
df_cars

starting_price  reviews_count  max_torque_nm  max_torque_rpm  \
0            399000             51           89.0            3500   
1            799000             86          136.8            4400   
2           1353000            242          300.0            2800   
3           1318000            313          450.0            2800   
4           1199000            107          400.0            2750   
..              ...            ...            ...             ...   
198          659000             35          500.0            5250   
199         1041000              3          400.0            4400   
200         1615000              2          350.0            2500   
201        21700000              9          800.0            4500   
202          391000              4           69.0            3500   

     max_power_bhp  max_power_rp  fuel_tank_capacity  no_cylinder  rating  \
0            65.71          5500                27.0            3     4.5   
1           101.65          6000                48.0            4     4.5   
2           130.00          3750                57.0            4     4.5   
3           182.38          3500                60.0            4     4.5   
4           172.45          3500                57.0            4     4.5   
..             ...           ...                 ...          ...     ...   
198         415.71          6750                 0.0            4     4.5   
199         254.79          5000                59.0            4     4.5   
200         167.67          3750                60.0            4     4.5   
201         591.39          6000                85.0            8     3.5   
202          47.33          6000                35.0            3     4.5   

     seating_capacity fuel_type  engine_displacement transmission_type  \
0                 5.0    Petrol                  998         Automatic   
1                 5.0    Petrol                 1462         Automatic   
2                 4.0    Diesel                 2184         Automatic   
3                 7.0    Diesel                 2198         Automatic   
4                 7.0    Diesel                 2198         Automatic   
..                ...       ...                  ...               ...   
198               5.0    Petrol                 1991         Automatic   
199               5.0    Petrol                 1998         Automatic   
200               7.0    Diesel                 1956            Manual   
201               5.0    Petrol                 3998         Automatic   
202               5.0    Petrol                  796            Manual   

     body_type  
0    Hatchback  
1          SUV  
2          SUV  
3          SUV  
4          SUV  
..         ...  
198  Hatchback  
199      Sedan  
200        SUV  
201        SUV  
202  Hatchback  

[203 rows x 14 columns]

### Removing Features Based on Incomplete or Missing Data

In our cleaning data chapter, we identified and populated null values where possible using a technique called imputation.

However this may not always be possible.

In [26]:
df_cars.isnull().sum()

starting_price         0
reviews_count          0
max_torque_nm          0
max_torque_rpm         0
max_power_bhp          0
max_power_rp           0
fuel_tank_capacity     0
no_cylinder            0
rating                 0
seating_capacity       1
fuel_type              0
engine_displacement    0
transmission_type      0
body_type              0
dtype: int64

If for whatever reason, a feature still has a high number of nulls, we may consider dropping the feature entirely. The decision to do this depends on how important or powerful the feature is likely to be in our model. We should also consider whether null values should be expected in the scenario.

- For example, in a feature containing Marriage_Date, we may expect a significant % of nulls.
- If an important feature from domain knowledge perspective has many nulls, we may ask the business to provide more data.

Note: There are no black and white rules for removing features.